File: csharp/Princeton-Algorithms1/Week01/Percolation/Percolation.cs
using System;

namespace Princeton_Algorithms1.Week01.Percolation
{
    /// <summary>
    ///     Percolation data type to model a percolation system.
    ///     Indexing is 0-based!
    /// </summary>
    /// <remarks>
    ///     See specification at:
    ///     https://coursera.cs.princeton.edu/algs4/assignments/percolation/specification.php
    /// </remarks>
    public class Percolation
    {
        private readonly IUnionFindImplementation _unionFind;
        private int _openSiteCounter;
        private readonly bool[,] siteStates;
        private readonly int _virtualTopIndex;
        private readonly int _virtualBottomIndex;

        public int Size { get; }

        /// <summary>
        ///     Initializes a new instance of the <see cref="Percolation"/> class,
        ///     by creating an n-by-n grid, with all sites initially blocked/full
        /// </summary>
        /// <param name="n">Size of the grid (will be n-by-n)</param>
        /// <param name="unionFind">Union find algorithm implementation that is NOT initialized (and its indexing is 0-based).</param>
        public Percolation(int n, IUnionFindImplementation unionFind)
        {
            if (n <= 0)
                throw new ArgumentOutOfRangeException(nameof(n), n, "Value cannot be less than 1.");

            Size = n;
            _unionFind = unionFind;
            _unionFind.Initialize(Size * Size + 2);
            siteStates = new bool[Size, Size];

            // Virtual top element
            _virtualTopIndex = Size * Size;
            InitializeVirtualTop();

            // Virtual bottom element
            _virtualBottomIndex = Size * Size + 1;
            InitializeVirtualBottom();
        }

        private void InitializeVirtualTop()
        {
            for (int i = 0; i < Size; i++)
            {
                int index = CalculateUnionFindIndex(0, i);
                _unionFind.Union(index, _virtualTopIndex);
            }
        }

        private void InitializeVirtualBottom()
        {
            for (int i = 0; i < Size; i++)
            {
                int index = CalculateUnionFindIndex(Size - 1, i);
                _unionFind.Union(index, _virtualBottomIndex);
            }
        }

        /// <summary>
        ///     Opens the given site
        /// </summary>
        public void Open(int row, int col)
        {
            CheckCoordinates(row, col);
            if (IsOpen(row, col))
                return;

            siteStates[row, col] = true;
            _openSiteCounter++;
            int index = CalculateUnionFindIndex(row, col);

            UnionIfOpen(row, col + 1);
            UnionIfOpen(row, col - 1);
            UnionIfOpen(row + 1, col);
            UnionIfOpen(row - 1, col);

            // ------------------------------------------
            void UnionIfOpen(int row, int col)
            {
                bool isOpen = IsOpenSafe(row, col);
                if (isOpen)
                    _unionFind.Union(index, CalculateUnionFindIndex(row, col));
            }
        }

        /// <summary>
        ///     Returns whether the given site is open.
        /// </summary>
        public bool IsOpen(int row, int col)
        {
            CheckCoordinates(row, col);
            return IsOpenNonChecking(row, col);
        }

        private bool IsOpenSafe(int row, int col)
        {
            if (IsInvalidCoordinatePart(row) || IsInvalidCoordinatePart(col))
                return false;
            return IsOpenNonChecking(row, col);
        }

        private bool IsOpenNonChecking(int row, int col)
        {
            return siteStates[row, col];
        }

        /// <summary>
        ///     Returns whether the given site is full (not open).
        /// </summary>
        public bool IsFull(int row, int col) => !IsOpen(row, col);

        /// <summary>
        ///     Returns the number of open sites
        /// </summary>
        public int NumberOfOpenSites() => _openSiteCounter;

        /// <summary>
        ///     Determines whether the system percolates.
        /// </summary>
        public bool Percolates()
        {
            if (Size == 1)
                return _openSiteCounter == 1;

            return _unionFind.Find(_virtualBottomIndex) == _unionFind.Find(_virtualTopIndex);
        }

        private void CheckCoordinates(int row, int col)
        {
            if (IsInvalidCoordinatePart(row))
                throw new ArgumentOutOfRangeException(nameof(row), row, "Value has to be between the interval [0;N[.");
            if (IsInvalidCoordinatePart(col))
                throw new ArgumentOutOfRangeException(nameof(col), col, "Value has to be between the interval [0;N[.");
        }

        private bool IsInvalidCoordinatePart(int x)
        {
            return x < 0 || x >= Size;
        }

        private int CalculateUnionFindIndex(int row, int col)
        {
            return row * Size + col;
        }
    }
}

File: java/percolation.princeton-algorithms1.arphox/src/Percolation.java
import edu.princeton.cs.algs4.WeightedQuickUnionUF;

public class Percolation {

    private final int size;
    private final boolean[][] siteStates;
    private final int virtualTopIndex;
    private final int virtualBottomIndex;
    private final WeightedQuickUnionUF unionFind;
    private final WeightedQuickUnionUF unionFindForFull;
    private int openSiteCounter;

    // creates n-by-n grid, with all sites initially blocked
    public Percolation(int n) {
        if (n <= 0)
            throw new IllegalArgumentException("Value of n has to be positive");
        size = n;
        unionFind = new WeightedQuickUnionUF(size * size + 2);
        unionFindForFull = new WeightedQuickUnionUF(size * size + 2);
        siteStates = new boolean[size][];
        initializeSiteStates();

        // Virtual top element
        virtualTopIndex = size * size;
        initializeVirtualTop();

        // Virtual bottom element
        virtualBottomIndex = size * size + 1;
        initializeVirtualBottom();
    }

    // test client (optional)
    public static void main(String[] args) {
        System.out.println("Percolation test");
    }

    private void initializeSiteStates() {
        for (int i = 0; i < size; i++) {
            siteStates[i] = new boolean[size];
        }
    }

    private void initializeVirtualTop() {
        for (int i = 0; i < size; i++) {
            int index = calculateUnionFindIndex(0, i);
            unionFind.union(index, virtualTopIndex);
            unionFindForFull.union(index, virtualTopIndex);
        }
    }

    private void initializeVirtualBottom() {
        for (int i = 0; i < size; i++) {
            int index = calculateUnionFindIndex(size - 1, i);
            unionFind.union(index, virtualBottomIndex);
        }
    }

    // opens the site (row, col) if it is not open already
    public void open(int row, int col) {
        checkCoordinates(row, col);
        if (isOpenSkipCheck(row, col))
            return;

        siteStates[row - 1][col - 1] = true;
        openSiteCounter++;
        int index = calculateUnionFindIndex(row - 1, col - 1);

        unionIfOpen(row, col + 1, index);
        unionIfOpen(row, col - 1, index);
        unionIfOpen(row + 1, col, index);
        unionIfOpen(row - 1, col, index);
    }

    private void unionIfOpen(int row, int col, int index) {
        boolean isOpen = isOpenNonThrowing(row, col);
        if (isOpen) {
            int indexOther = calculateUnionFindIndex(row - 1, col - 1);
            unionFind.union(index, indexOther);
            unionFindForFull.union(index, indexOther);
        }
    }

    // is the site (row, col) open?
    public boolean isOpen(int row, int col) {
        checkCoordinates(row, col);
        return isOpenSkipCheck(row, col);
    }

    // is the site (row, col) full?
    public boolean isFull(int row, int col) {
        checkCoordinates(row, col);
        boolean isOpen = isOpen(row, col);
        if (!isOpen)
            return false;

        int index = calculateUnionFindIndex(row - 1, col - 1);
        return unionFindForFull.find(index) == unionFindForFull.find(virtualTopIndex);
    }

    // returns the number of open sites
    public int numberOfOpenSites() {
        return openSiteCounter;
    }

    // does the system percolate?
    public boolean percolates() {
        if (size == 1)
            return openSiteCounter == 1;

        return unionFind.find(virtualBottomIndex) == unionFind.find(virtualTopIndex);
    }

    private boolean isInvalidCoordinatePart(int x) {
        return x < 1 || x > size;
    }

    private int calculateUnionFindIndex(int rowZeroBased, int colZeroBased) {
        return rowZeroBased * size + colZeroBased;
    }

    private boolean isOpenSkipCheck(int row, int col) {
        return siteStates[row - 1][col - 1];
    }

    private boolean isOpenNonThrowing(int row, int col) {
        if (isInvalidCoordinatePart(row) || isInvalidCoordinatePart(col))
            return false;
        return isOpenSkipCheck(row, col);
    }

    private void checkCoordinates(int row, int col) {
        if (isInvalidCoordinatePart(row))
            throw new IllegalArgumentException("Value of row has to be between the interval [1;N].");
        if (isInvalidCoordinatePart(col))
            throw new IllegalArgumentException("Value of col has to be between the interval [1;N].");
    }
}